Dem force chain ghost (#1172)
chaos-polymtl/lethe: doc/source/parameters/dem/post-processing.rst

Post-processing
===============

> Lethe-DEM has built-in post-processing capabilities. All post-processing results are written as VTU, PTVU, and/or PVD files. 

Reviewer: ```suggestion
Lethe has built-in post-processing capabilities. All post-processing results are written as VTU, PTVU, and/or PVD files. 
```

Reviewer: I know it was written Lethe-DEM before, but a year ago we started to replace every "Lethe-DEM" by Lethe.

The post-processing subsection of the ``.prm`` file is according to the following example:

.. code-block:: text

 subsection post-processing
  # Enable output of grid, granular temperature, and particles' average velocity
  set Lagrangian post-processing = false
  # Enable output of force chains
  set force chains = false
 end

.. note::

>  By default, post-processing are set to ``false``.

Reviewer 2: ```suggestion
 By default, post-processing options are set to ``false`` to reduce the number of files generated.
```

>  By default, post-processing options are set to ``false`` to reduce te number of files generated.

Reviewer 2: ```suggestion
 By default, post-processing options are set to ``false`` to reduce the number of files generated.
```
```suggestion
 By default, post-processing options are set to ``false`` to reduce te number of files generated.
```

--------------------------
Lagrangian post-processing
--------------------------

> The ``Lagrangian post-processing`` enables VTU files output generated with the same ``output frequency`` as the particles' results (see the :doc:`../dem/simulation_control` section of this guide). The name of the generated files are the same the one chosen for the particles' results plus the suffix ``-postprocess_data``. The files include:

Reviewer: ```suggestion
The ``Lagrangian post-processing`` enables VTU files output generated with the same ``output frequency`` as the particles' results (see the :doc:`../dem/simulation_control` section of this guide). The name of the generated files are set by the ``output name`` parameter in the simulation control section plus the suffix ``-postprocess_data``. The files include:
```

Reviewer: Got two extra comments/questions 

- It think the suffix for the post-processing lagrangian should be different since we have more than one post processing tool (force chain and this).

- The ``output name`` parameter is not documented in the simulation control page. (It could be fix in this PR since it is a tiny change, but we can fix it in an other one) (@blaisb should I open a issue for this?)

Reviewer 2: Yeah you can open an issue and the first one with time can add a PR to add it to the documentation

Author: For the force chains, I'll change the name of the files for the output name plus the suffix "-force_chains". Do you want me to modify the names for post-processing lagrangian files in this PR?

> The ``Lagrangian post-processing`` enables VTU files output generated with the same ``output frequency`` as the particles' results (see the :doc:`../dem/simulation_control` section of this guide). The name of the generated files are set by the ``output name`` parameter in the simulation_control section plus the suffix ``-postprocess_data``. The files include:

Reviewer 2: ```suggestion
The ``Lagrangian post-processing`` enables VTU files output generated with the same ``output frequency`` as the particles' results (see the :doc:`../dem/simulation_control` section of this guide). The name of the generated files is set by the ``output name`` parameter in the simulation_control section plus the suffix ``-postprocess_data``. The files include:
```

The name is singular here, so the name is set, even if its files.

.. important::
 The particles' average velocity and the granular temperature are cell based results calculated instantaneously at each outputting time-step. This means that they are only instantaneous spatial averages, i.e, DO NOT correspond to temporal averages measured between the outputs. For a higher resolution of the results in time, we recommend the use of a lower ``output frequency`` (:doc:`../dem/simulation_control`).

----------------------------
Force chains post-processing
----------------------------

> The ``force chains`` feature enables VTU, PTVU, and PVD files to visualize force chains between particles. Generated files's names begin with the prefix ``force_chains-``. The files include:

Reviewer 2: This one uses a suffix and the previous one (see above) uses a prefix. Maybe we should uniformize this to have only suffixes?

Reviewer: agree. 

Both sufix should be set by the ``output name`` parameter followed by either "force_chains" or "lagrangian_post_processing" (or something else)

* Lines (use wireframe display with paraview) to represent forces. 

* The force values for each line.

> .. important::

Reviewer 2: ```suggestion
.. warning::
```

Reviewer: I don't thing it's super expensive. Forces are being computed at every time step. But maybe I'm wrong.

Reviewer 2: No it's not that expensive that's for sure, I think making the triangulation and writing it is more expensive than the force calculation. I think that adding an image and maybe even a link to the youtube video would be a good idea

.. important::
 The particles' average velocity and the granular temperature are cell based results calculated instantaneously at each outputting time-step. This means that they are only instantaneous spatial averages, i.e, DO NOT correspond to temporal averages measured between the outputs. For a higher resolution of the results in time, we recommend the use of a lower ``output frequency`` (:doc:`../dem/simulation_control`).

----------------------------
Force chains post-processing
----------------------------

> The ``force chains`` feature enables VTU, PTVU, and PVD files to visualize force chains between particles. The name of the generated files are set by the ``output name`` parameter in the simulation_control section plus the suffix ``-force_chains``. The files include:

Reviewer 2: ```suggestion
The ``force chains`` feature enables VTU, PTVU, and PVD files to visualize force chains between particles. The name of the generated files is set by the ``output name`` parameter in the simulation_control section plus the suffix ``-force_chains``. The files include:
```

> * Lines to represent forces between particles in contact (use wireframe display with paraview).

Reviewer: ```suggestion
* Lines to represent normal forces between particles in contact (use wireframe display with paraview).
```

> * Lines to represent forces between particles in contact. (use wireframe display with paraview)

Reviewer 2: ```suggestion
* Lines to represent forces between particles in contact (use wireframe display with paraview).
```

* The force value for each line.

.. warning::

>  The force chains feature is not implemented for periodic domains. 

Reviewer: ```suggestion
 The force chains feature is not implemented for contact occurring through periodic boundaries. 
```

Reviewer: The feature still works for periodic domains.

The following video shows force chains for the `Rectangular Hopper example <https://chaos-polymtl.github.io/lethe/documentation/examples/dem/rectangular-hopper/rectangular-hopper.html>`_.

 .. raw:: html

>     <iframe width="600" height="303" src="https://youtu.be/XrXXCz00Yjk?si=6T9_PhBDX3SkUsne" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Reviewer 3: Am I the only one that sees that? The link is working well, but this is the only embedded YT video that shows me this. Maybe it is because it's not on the website (master) yet?
![image](https://github.com/chaos-polymtl/lethe/assets/71884806/1c37a1c1-bbee-496a-a80e-15edfc6d0565)

Author: Same here. The link is working but I need to open it on another page.

Reviewer 2: I think it's because you did not put the embed youtube video link
It's not the same as directly linking to the youtube video, you need to grab the embedded URL. You can check on youtube directly how to get this URL, but it should have a different prefix.
